### PR TITLE
Allow swarm join with `--availability=drain`

### LINKED ISSUE #24596: Cannot join Swarm as Manager only
When joining a swarm, it is not possible to join as manager only. This behavior is not desirable in scenarios where we want dedicated masters.

**Output of `docker version`:**

```
Client:
 Version:      1.12.0-rc3
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   91e29e8
 Built:        Sat Jul  2 00:40:04 2016
 OS/Arch:      linux/amd64
 Experimental: true

Server:
 Version:      1.12.0-rc3
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   91e29e8
 Built:        Sat Jul  2 00:40:04 2016
 OS/Arch:      linux/amd64
 Experimental: true

```

### PR DESCRIPTION
**Pull request by yongtang**
**\- What I did**

This fix tries to address the issue raised in #24596 where it was not possible to join as manager only (`--availability=drain`).

**The related PR in swarmkit is:**
https://github.com/docker/swarmkit/pull/1271

**\- How I did it**

This fix adds a new flag `--availability` to `swarm join`.

**\- How to verify it**

An integration test has been added.

**\- Description for the changelog**

Add a new flag `--availability` to `swarm join`.

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes #24596.

**The related PR in swarmkit is:**

https://github.com/docker/swarmkit/pull/1271

### CODE REVIEW

**Comment by thaJeztah** [comment]
I'm +1 on design, I think it makes sense to haven an option to set the initial availability of a node

**Comment by thaJeztah** [comment]
Tentatively adding this to the milestone (because the related issue is), but it's no show-stopper

File: vendor/src/github.com/docker/swarmkit/api/ca.proto
Comment (aaronlehmann) [inline vendor/src/github.com/docker/swarmkit/api/ca.proto]: This will be ignored for a renewal. I think that's okay, but we need to make it clear in the comment.

I'm not sure we want to add `availability` to the top level of `IssueNodeCertificateRequest`. Either there should be a nested messages with parameters for the new node object, or we should pass a `NodeSpec` and only honor certain fields from it (because not everything in a `NodeSpec` is under the node's control).

@aluzzardi @stevvooe any thoughts?

Comment (aluzzardi) [inline vendor/src/github.com/docker/swarmkit/api/ca.proto]: I don't like much `availability` leaking to the `NodeCA`.

Also - if we go down this path, we should probably take an initial NodeSpec rather than a bunch of fields

File: vendor/src/github.com/docker/engine-api/types/swarm/swarm.go
Comment (stevvooe) [inline vendor/src/github.com/docker/engine-api/types/swarm/swarm.go]: Surely, we have a type for this.

File: vendor/src/github.com/docker/swarmkit/api/ca.proto
Comment (stevvooe) [inline vendor/src/github.com/docker/swarmkit/api/ca.proto]: This really isn't the right place for it.

Agree with the concept of providing a `NodeSpec` at join time.

**Comment by stevvooe** [comment]
In general, this is a good feature, but the implementation needs some design work.

Do we have something open in swarmkit?

File: vendor/src/github.com/docker/engine-api/types/swarm/swarm.go
Comment (aaronlehmann) [inline vendor/src/github.com/docker/engine-api/types/swarm/swarm.go]: Yes, the type is called `NodeAvailability`.

File: vendor/src/github.com/docker/swarmkit/api/ca.proto
Comment (aaronlehmann) [inline vendor/src/github.com/docker/swarmkit/api/ca.proto]: @yongtang: It sounds like the consensus is that we should provide a `NodeSpec` instead of just an availability setting. But as I mentioned earlier, the node CA should only use certain fields from this spec when it creates the node. There are things like node labels that are controlled from elsewhere.

**Comment by yongtang** [comment]
@stevvooe @aaronlehmann @aluzzardi @vdemeester @thaJeztah Thanks a lot for the review and help!

I was looking for feedbacks on this PR before, so I haven't created the pull request on swarmkit side yet.

Let me rework on this pull request based on the comments. Will work on `NodeSpec` and update the pull request soon.

**Comment by icecrime** [comment]
Thanks @yongtang! In the meantime, I'll remove the milestone as it won't make it in time for 1.12.0 (which is ok).

**Comment by cpuguy83** [comment]
ping @stevvooe

File: vendor/src/github.com/docker/engine-api/types/swarm/swarm.go
Comment (stevvooe) [inline vendor/src/github.com/docker/engine-api/types/swarm/swarm.go]: Bump. Please use the type.

**Comment by stevvooe** [comment]
@cpuguy83 I am not necessarily on board with using `NodeSpec` here, since we don't honer the other aspects of the structure, such as node role.

File: vendor/src/github.com/docker/engine-api/types/swarm/swarm.go
Comment (yongtang) [inline vendor/src/github.com/docker/engine-api/types/swarm/swarm.go]: Thanks @stevvooe. The type has been changed to `NodeAvailability`.

File: daemon/cluster/cluster.go
Comment (stevvooe) [inline daemon/cluster/cluster.go]: This is way too many string arguments in a row.

**Comment by cpuguy83** [comment]
ping @yongtang

File: daemon/cluster/cluster.go
Comment (yongtang) [inline daemon/cluster/cluster.go]: Thanks @stevvooe. The pull request has been updated with `localAddr`, `listenAddr`, and `advertiseAddr` into one consolidated `nodeAddr`. Please let me know if there are any issues.

**Comment by yongtang** [comment]
Thanks @cpuguy83. The pull request has been rebased and updated. Please let me know if there are any issues.

**Comment by thaJeztah** [comment]
Sorry @yongtang this needs another rebase

ping @stevvooe is this ok to be moved to code review?

**Comment by stevvooe** [comment]
This was my last comment here:

> @cpuguy83 I am not necessarily on board with using NodeSpec here, since we don't honer the other aspects of the structure, such as node role.

If we can reconcile this, then we can probably move along.

**Comment by aaronlehmann** [comment]
To unblock this, maybe we can go back to the approach of not using `NodeSpec`.

Using `NodeSpec` was my suggestion. I'm sorry it delayed this so much.

**Comment by stevvooe** [comment]
@aaronlehmann I have no problem using `NodeSpec`, as long as we don't cherry pick parameters. I think the problem is that `NodeSpec` can set things like the name, which should be under control of the cluster administrator, or membership, which is the purview of the managers. `NodeSpec` is supposed to be the user input for a node.

Let's just make this a field value for "request availability", ultimately leaving the policy to the orchestrator as to whether it is honored or not.

Do we also incorporate this into the join token?

**Comment by aaronlehmann** [comment]
> Let's just make this a field value for "request availability", ultimately leaving the policy to the orchestrator as to whether it is honored or not.

Agreed.

> Do we also incorporate this into the join token?

My feeling is that we shouldn't, to prevent the setting from being opaque and to limit the number of permutations on the join token from exploding, but it's an interesting idea.

**Comment by thaJeztah** [comment]
@yongtang do you have enough information to work on this?

**Comment by yongtang** [comment]
Thanks @stevvooe @aaronlehmann @thaJeztah. Let me take a look and update the PR accordingly.

**Comment by yongtang** [comment]
@stevvooe @aaronlehmann @thaJeztah The PR has been updated. Please take a look and let me know if there are any issues.

File: cli/command/swarm/join.go
Comment (aaronlehmann) [inline cli/command/swarm/join.go]: Should this have a default value of "active"?

File: daemon/cluster/cluster.go
Comment (aaronlehmann) [inline daemon/cluster/cluster.go]: By default, this should be an empty string. We should only send an availability value if the user specifically chooses one with the `--availability` flag.

File: vendor/src/github.com/docker/swarmkit/ca/server.go
Comment (aaronlehmann) [inline vendor/src/github.com/docker/swarmkit/ca/server.go]: Before using this in `NodeSpec`, check that it is non-empty. Old engines won't include this in the request (and I think new ones shouldn't as well, unless they need to request a specific availability value).

File: daemon/cluster/cluster.go
Comment (stevvooe) [inline daemon/cluster/cluster.go]: Why is this type exported?

File: daemon/cluster/cluster.go
Comment (aaronlehmann) [inline daemon/cluster/cluster.go]: Happy to examine it.

- LocalAddr: The node's local address. Needed for setting up a tunnel.
- RemoteAddr: The address passed of an existing node passed to `swarm join`.
- ListenAddr: The bind address. This is usually 0.0.0.0 or `::`, so it is not redundant with `LocalAddr`.
- AdvertiseAddr: Externally-reachable address that's advertised to other nodes. Needed in one-to-one NAT scenarios.

File: cli/command/swarm/join.go
Comment (yongtang) [inline cli/command/swarm/join.go]: Thanks @aaronlehmann. The PR has been updated with default value `active`.

File: cli/command/swarm/join.go
Comment (yongtang) [inline cli/command/swarm/join.go]: Thanks. Done.

File: daemon/cluster/cluster.go
Comment (yongtang) [inline daemon/cluster/cluster.go]: Thanks. Done.

File: vendor/src/github.com/docker/swarmkit/ca/server.go
Comment (yongtang) [inline vendor/src/github.com/docker/swarmkit/ca/server.go]: @aaronlehmann The `request.Availability` is defined as 
```
type NodeSpec_Availability int32
```
It seems the default value will be `0` so not sure we could check if this field is set or not.

File: daemon/cluster/cluster.go
Comment (yongtang) [inline daemon/cluster/cluster.go]: Thanks @stevvooe. The changes related to this part has already been done in PR #27698. I just removed related code accordingly.

**Comment by yongtang** [comment]
Thanks @stevvooe @aaronlehmann @thaJeztah for the review. The PR has been updated and the comments have been addressed. (Other than the none-empty check of `request.Availability`).

Please take a look and let me know if there are any issues.

File: vendor/src/github.com/docker/swarmkit/ca/server.go
Comment (aaronlehmann) [inline vendor/src/github.com/docker/swarmkit/ca/server.go]: Sorry, for some reason I thought this was a string. In this case, I think it makes sense to make it a pointer, and keep the value `nil` if there the user did not override it. On the gRPC side, this means splitting it out into a separate message.

Comment (yongtang) [inline vendor/src/github.com/docker/swarmkit/ca/server.go]: @aaronlehmann Currently the definition of `Availability` is embedded inside the `NodeSpec`:
```
message NodeSpec {
        Annotations annotations = 1 [(gogoproto.nullable) = false];

        enum Membership {
                option (gogoproto.goproto_enum_prefix) = false;

                PENDING = 0 [(gogoproto.enumvalue_customname) = "NodeMembershipPending"];
                ACCEPTED = 1 [(gogoproto.enumvalue_customname) = "NodeMembershipAccepted"];
        }

        enum Availability {
                option (gogoproto.goproto_enum_prefix) = false;

                // Active nodes.
                ACTIVE = 0 [(gogoproto.enumvalue_customname) = "NodeAvailabilityActive"];
                ...
        }
```

Do we want to move `Availability` outside of `NodeSpec` (change `NodeSpec`)? Or we could spin off a new message (maybe `NodeAvailability`?) to include `NodeSpec.Availability`, and keep current `NodeSpec` message intact, like the following?
```
message NodeAvailability {

        NodeSpec.Availability availability = 1;

}
```

Comment (aaronlehmann) [inline vendor/src/github.com/docker/swarmkit/ca/server.go]: I don't think we should move `Availability` out of `NodeSpec`.

What I was thinking of was having a wrapper mesasge instead of using `NodeSpec.Availability` directly in `IssueNodeCertificateRequest`. Something like this:

```
message NodeAvailability {
      NodeSpec.Availability availability = 1;
}
```

and in `IssueNodeCertificateRequest`:

```
...
      // Availability allows a user to control the current scheduling status of a node
      NodeAvailability availability = 4;
```

This means `availability` can be nil. If the user does not specify an availability, we can choose a default on the server side.

**Comment by yongtang** [comment]
Also, the related PR in swarmkit (https://github.com/docker/swarmkit/pull/1271) has been updated as well. Please take a look and let me know for any additional issues.

File: cli/command/swarm/join.go
Comment (aaronlehmann) [inline cli/command/swarm/join.go]: It might make sense to do the `swarm.NodeAvailability` cast here, to avoid the casts below.

File: cli/command/swarm/join.go
Comment (aaronlehmann) [inline cli/command/swarm/join.go]: Do you think we should support this flag on `docker swarm init` too?

File: cli/command/swarm/join.go
Comment (yongtang) [inline cli/command/swarm/join.go]: Thanks @aaronlehmann. Done.

File: cli/command/swarm/join.go
Comment (yongtang) [inline cli/command/swarm/join.go]: Thanks. Done.

File: cli/command/swarm/join.go
Comment (aaronlehmann) [inline cli/command/swarm/join.go]: It's probably easiest to do it in this PR.

**Comment by aaronlehmann** [comment]
LGTM (after swarmkit PR is merged)

Thanks for sticking with this!

**Comment by yongtang** [comment]
Thanks @aaronlehmann for the review. The windowsRS1 is failing but that seems to be an Jenkins issue.

**Comment by yongtang** [comment]
Thanks all for the review. The PR has been updated with vendoring of swarmkit done. Now all tests passes. Please take a look and let me know if there are additional issues.

**Comment by aaronlehmann** [comment]
LGTM after rebase

**Comment by yongtang** [comment]
@aaronlehmann Rebased. Thanks.

**Comment by cpuguy83** [comment]
What's the status here?

**Comment by aaronlehmann** [comment]
The swarmkit-side PR was merged, so this is most of the way there. Sadly, it needs yet another rebase. It still looks good to me. Needs a second LGTM.

**Comment by yongtang** [comment]
@cpuguy83 @aaronlehmann The PR has been rebased.

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by aaronlehmann** [comment]
@thaJeztah: What are the next steps to get this merged? It's not an urgent PR, but I feel bad about it being open so long, and asking for rebases every few weeks.

**Comment by thaJeztah** [comment]
oh, my bad, lost track of this one 😞

**Review by thaJeztah — CHANGES_REQUESTED** [review]
left some suggestions

File: docs/reference/commandline/swarm_join.md
Comment (yongtang) [inline docs/reference/commandline/swarm_join.md]: Thanks. Done.

File: docs/reference/commandline/swarm_join.md
Comment (yongtang) [inline docs/reference/commandline/swarm_join.md]: Thanks. Done.

File: docs/reference/commandline/swarm_init.md
Comment (yongtang) [inline docs/reference/commandline/swarm_init.md]: Done.

File: docs/reference/commandline/swarm_init.md
Comment (yongtang) [inline docs/reference/commandline/swarm_init.md]: Done.

**Comment by yongtang** [comment]
Thanks @aaronlehmann @thaJeztah the PR has been updated.

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by genki** [comment]
Is there a way to become availability=active from worker node itself?

**Pull request merged**
